### PR TITLE
improve strictness of spree_line_items

### PR DESCRIPTION
**Pull request by BenMorganIO**
## Summary

The goal of this migration is to add `NOT NULL` where applicable and to add foreign key constraints. The foreign keys on_delete are copies of what the `dependent` option would likely replicate.

## Checklist

Check out our [PR guidelines](https://github.com/solidusio/.github/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] [I have used clear, explanatory commit messages](https://github.com/solidusio/.github/blob/main/CONTRIBUTING.md#writing-good-commit-messages).

The following are not always needed:

- 📖 I have updated the README to account for my changes.
- 📑 I have documented new code [with YARD](https://www.rubydoc.info/gems/yard/file/docs/Tags.md).
- 🛣️ I have opened a PR to update the [guides](https://github.com/solidusio/edgeguides).
- ✅ I have added automated tests to cover my changes.
- 📸 I have attached screenshots to demo visual changes.

### CODE REVIEW

**Review by tvdeyen — COMMENTED** [review]
I like that change, but we should at least give owners of very old stores some hints, what to do if the migration fail for them (mostly concerned about deleted variants)

File: core/db/migrate/20230328042835_improve_strictness_of_spree_line_items.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
+class ImproveStrictnessOfSpreeLineItems < ActiveRecord::Migration[7.0]
```
Comment (kennyadsl) [inline core/db/migrate/20230328042835_improve_strictness_of_spree_line_items.rb:1]:
```suggestion
class ImproveStrictnessOfSpreeLineItems < ActiveRecord::Migration[5.2]
```

Comment (kennyadsl) [reply core/db/migrate/20230328042835_improve_strictness_of_spree_line_items.rb:1]: ☝️ this will likely make the test pass on all Rails versions.

Comment (tvdeyen) [reply core/db/migrate/20230328042835_improve_strictness_of_spree_line_items.rb:1]: I know that we do this in all our migrations, but I want to suggest that we do something like this

```rb
class ImproveStrictnessOfSpreeLineItems < ActiveRecord::Migration[ActiveRecord::Migration.current_version]
```

to make sure that new Stores get all features (like `t.bigint`, `NOT NULL` on timestamps, etc.) for later Rails versions.

To make it easier we could introduce a `Spree::Migration` base class.

Comment (kennyadsl) [reply core/db/migrate/20230328042835_improve_strictness_of_spree_line_items.rb:1]: The only problem with this approach is that we won't be sure of how the database of users looks like anymore. A middle-ground could be using the last version of Rails that we support (currently 6.0) for new migrations?

Comment (kennyadsl) [reply core/db/migrate/20230328042835_improve_strictness_of_spree_line_items.rb:1]: I mean, but correct me if I'm wrong, we won't know which migrations have been executed on different stores, depending on the version of Rails they are using. This way, we could have a future migration (or application code) that will need conditional paths depending on the Rails version used when the migration has been executed.

Comment (BenMorganIO) [reply core/db/migrate/20230328042835_improve_strictness_of_spree_line_items.rb:1]: We could just do a large migration and set all the timestamps as `not null`. That seems to be the appropriate solution. Then we could do another migration later where we make all the primary/foreign keys bigints.

**Comment by BenMorganIO** [comment]
@tvdeyen @kennyadsl I think this should be reserved for a major version number change and it can be highlighted in the changelog. I think we should create several migrations that focus on what we want to improve:

- `NOT NULL` changes only on the timestamps
- `NOT NULL` changes on non-timestamp columns
- `FOREIGN KEY CONSTRAINT` for all the missing relations
- `bigint` for primary key/foreign key columns

File: core/db/migrate/20230328042835_improve_strictness_of_spree_line_items.rb (hunk 1 above)
Comment (elia) [reply core/db/migrate/20230328042835_improve_strictness_of_spree_line_items.rb:1]: This should be fine after merging https://github.com/solidusio/solidus/pull/5012

**Comment by waiting-for-dev** [comment]
We should write very detailed release instructions about how to check if a store will be affected by the change. Something like we did in https://github.com/solidusio/solidus/pull/4851. The text would be copied to the corresponding upgrade instructions on our guides.

**Comment by waiting-for-dev** [comment]
Rebasing from master and marking a test as pending, as we can't simulate an invalid line item with this change

**Comment by kennyadsl** [comment]
Closing this for the same reasons we did close https://github.com/solidusio/solidus/pull/2036. Thanks again for the contribution @BenMorganIO; I'm sure it will come the day we will be able to improve the strictness of our db, and that day we will resume this PR considering your proposal solution. 

🙏

**Pull request closed without merging**